File: greedy/candy.js
/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function(ratings) {
 let n = ratings.length;
 let count = 0;
 if(ratings[0]>ratings[1]) {
    count += 2;
 }
 if(ratings[0]<=ratings[1]) {
    count += 1;
 }
 if(ratings[n-1]>ratings[n-2]) count+=2;
 if(ratings[n-1]<=ratings[n-2]) count+=1;
 for(let i =1;i<n-1;i++){
    if(ratings[i]>=ratings[i-1] && ratings[i]>=ratings[i+1]){
        count+=2;
    }
    else{
        count += 1;
    }
 }  
 return count; 
};

File: greedy/candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        int sum = 1;
        int i = 1;

        while (i < n) {
            int peak = 0;
            int down = 0;

            while (i < n && ratings[i] > ratings[i - 1]) {
                peak++;
                sum += peak + 1;
                i++;
            }

            while (i < n && ratings[i] < ratings[i - 1]) {
                down++;
                sum += down;
                i++;
            }

            if (down > 0) {
                sum -= min(peak, down);
            }

            if (i < n && ratings[i] == ratings[i - 1]) {
                sum++;
                i++;
            }
        }

        return sum;
    }
};